### ISSUE TITLE
Negative int offset on element function throws goroutine panic

### ISSUE THREAD

**Issue by enichols-insight** [hashicorp/terraform#25775]
### Terraform Version
```
0.12.29
```

### Terraform Configuration Files
```terraform
resource "azurerm_subnet" "example" {
  for_each             = { for suite, config in local.suites : suite => config if(config.enable_az_backup && lookup(config, "legacy", false)) }
  name                 = each.value.vnet_name
  resource_group_name  = data.terraform_remote_state.common-services.outputs.vnet[element(split("-", each.value.vnet_name), -1)].vnet_rg
  virtual_network_name = data.terraform_remote_state.common-services.outputs.vnet[element(split("-", each.value.vnet_name), -1)].name
  address_prefixes     = data.terraform_remote_state.common-services.outputs.vnet[element(split("-", each.value.vnet_name), -1)].address_spaces
}
```

### Crash Output
```
Error: Error in function call

  on legacy_managed_rgs.tf line 31, in resource "azurerm_subnet" "example":
  31:   address_prefixes     = data.terraform_remote_state.common-services.outputs.vnet[element(split("-", each.value.vnet_name), -1)].address_spaces
    |----------------
    | each.value.vnet_name is "vnt-dev001-use2-dev-002"

Call to function "element" failed: panic in function implementation: element
key for list must be non-negative integer
goroutine 175 [running]:
runtime/debug.Stack(0xc000c32e40, 0x27231c0, 0x2fec830)
        /usr/local/go/src/runtime/debug/stack.go:24 +0x9d
github.com/zclconf/go-cty/cty/function.errorForPanic(...)
        /go/pkg/mod/github.com/zclconf/go-cty@v1.2.1/cty/function/error.go:44
github.com/zclconf/go-cty/cty/function.Function.Call.func1(0xc000c332a0,
0xc000c332c0)
        /go/pkg/mod/github.com/zclconf/go-cty@v1.2.1/cty/function/function.go:287
+0x8f
panic(0x27231c0, 0x2fec830)
        /usr/local/go/src/runtime/panic.go:522 +0x1b5
github.com/zclconf/go-cty/cty.Value.Index(0x3068dc0, 0xc0015864d0, 0x26fcce0,
0xc000221880, 0x3068d40, 0xc0000ce36f, 0x2af2b60, 0xc0006b6450, 0x5,
0x3068d40, ...)
        /go/pkg/mod/github.com/zclconf/go-cty@v1.2.1/cty/value_ops.go:762 +0xd63
github.com/hashicorp/terraform/lang/funcs.glob..func6(0xc0001bb500, 0x2, 0x2,
0x3068d40, 0xc0000ce380, 0xc0000ce380, 0x0, 0x0, 0x2b12be5, 0x3, ...)
        /home/circleci/project/project/lang/funcs/collection.go:75 +0x34f
github.com/zclconf/go-cty/cty/function.Function.Call(0xc000576150,
0xc0001bb500, 0x2, 0x2, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
        /go/pkg/mod/github.com/zclconf/go-cty@v1.2.1/cty/function/function.go:291
+0x51c
github.com/hashicorp/hcl/v2/hclsyntax.(*FunctionCallExpr).Value(0xc000186f00,
0xc000e5ae20, 0x3068d80, 0x4280e00, 0x27ea300, 0x42807a0, 0xc0000ba560, 0x1,
0x1)
        /go/pkg/mod/github.com/hashicorp/hcl/v2@v2.3.0/hclsyntax/expression.go:398
+0x10fd
github.com/hashicorp/hcl/v2/hclsyntax.(*IndexExpr).Value(0xc000581260,
0xc000e5ae20, 0xc0000ce380, 0x4, 0x23538a2300000003, 0xc000c34668, 0x0, 0x0,
0x0)
        /go/pkg/mod/github.com/hashicorp/hcl/v2@v2.3.0/hclsyntax/expression.go:644
+0xda
github.com/hashicorp/hcl/v2/hclsyntax.(*RelativeTraversalExpr).Value(0xc0002e38f0,
0xc000e5ae20, 0x0, 0x10, 0xc000b46e58, 0x1, 0x0, 0x0, 0x0)
        /go/pkg/mod/github.com/hashicorp/hcl/v2@v2.3.0/hclsyntax/expression.go:140
+0x55
github.com/hashicorp/hcl/v2/hcldec.(*AttrSpec).decode(0xc000e94240,
0xc00043a720, 0x0, 0x0, 0x0, 0xc000e5ae20, 0x0, 0x0, 0x0, 0x0, ...)
        /go/pkg/mod/github.com/hashicorp/hcl/v2@v2.3.0/hcldec/spec.go:205 +0x2a9
github.com/hashicorp/hcl/v2/hcldec.ObjectSpec.decode(0xc000e941e0,
0xc00043a720, 0x0, 0x0, 0x0, 0xc000e5ae20, 0xc000c34ca8, 0xc000705840, 0x20,
0x20, ...)
        /go/pkg/mod/github.com/hashicorp/hcl/v2@v2.3.0/hcldec/spec.go:79 +0x239
github.com/hashicorp/hcl/v2/hcldec.decode(0x3068100, 0xc0002214a0, 0x0, 0x0,
0x0, 0xc000e5ae20, 0x3068b00, 0xc000e941e0, 0x0, 0x0, ...)
        /go/pkg/mod/github.com/hashicorp/hcl/v2@v2.3.0/hcldec/decode.go:21 +0x118
github.com/hashicorp/hcl/v2/hcldec.Decode(...)
        /go/pkg/mod/github.com/hashicorp/hcl/v2@v2.3.0/hcldec/public.go:15
github.com/hashicorp/terraform/lang.(*Scope).EvalBlock(0xc000c19cc0,
0x3067940, 0xc000e941b0, 0xc0007057c0, 0x1, 0x1, 0x0, 0x0, 0x0, 0xc000ae3558,
...)
        /home/circleci/project/project/lang/eval.go:68 +0x2aa
github.com/hashicorp/terraform/terraform.(*BuiltinEvalContext).EvaluateBlock(0xc000faa340,
0x3067940, 0xc000e941b0, 0xc0007057c0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /home/circleci/project/project/terraform/eval_context_builtin.go:272 +0x2b6
github.com/hashicorp/terraform/terraform.(*EvalDiff).Eval(0xc0005c3180,
0x309f040, 0xc000faa340, 0x2, 0x2, 0x0, 0x0)
        /home/circleci/project/project/terraform/eval_diff.go:137 +0x273
github.com/hashicorp/terraform/terraform.EvalRaw(0x30235e0, 0xc0005c3180,
0x309f040, 0xc000faa340, 0x0, 0x0, 0x0, 0x0)
        /home/circleci/project/project/terraform/eval.go:57 +0x131
github.com/hashicorp/terraform/terraform.(*EvalSequence).Eval(0xc0007fbe20,
0x309f040, 0xc000faa340, 0x2, 0x2, 0x1b3db95, 0x3023a00)
        /home/circleci/project/project/terraform/eval_sequence.go:20 +0xfd
github.com/hashicorp/terraform/terraform.EvalRaw(0x3023920, 0xc0007fbe20,
0x309f040, 0xc000faa340, 0x27bfa80, 0x4134a64, 0x27231c0, 0xc0005688d0)
        /home/circleci/project/project/terraform/eval.go:57 +0x131
github.com/hashicorp/terraform/terraform.Eval(0x3023920, 0xc0007fbe20,
0x309f040, 0xc000faa340, 0xc0007fbe20, 0x3023920, 0xc0007fbe20, 0x0)
        /home/circleci/project/project/terraform/eval.go:35 +0x4d
github.com/hashicorp/terraform/terraform.(*Graph).walk.func1(0x2a797e0,
0xc0011ce160, 0x0, 0x0, 0x0)
        /home/circleci/project/project/terraform/graph.go:90 +0xf40
github.com/hashicorp/terraform/dag.(*Walker).walkVertex(0xc000148180,
0x2a797e0, 0xc0011ce160, 0xc000cc63c0)
        /home/circleci/project/project/dag/walk.go:392 +0x353
created by github.com/hashicorp/terraform/dag.(*Walker).Update
        /home/circleci/project/project/dag/walk.go:314 +0xa9b
.
```

### Expected Behavior
I'd expect TF to gracefully through just the error with the use of a non-negative INT for the element function rather than the Goroutine panic.

### Actual Behavior
Error + Goroutine panic

### Steps to Reproduce
Use of a negative offset for the element function

### Additional Context
No CI. Nothing unusual here. Low priority issue here but probably worth cleaning up if possible.

**Comment by danieldreier** [comment]
Thanks for reporting this! I've confirmed that this happened on 0.13.0 and put together a simplified reproduction case:

```hcl
output "example" {
  value = element(split("-", "ruh-roh"), -1)
}
```

**Comment by mildwonkey** [comment]
The linked PR fixes this issue, but unfortunately there are just enough other changes to the upstream go-cty library that I'm not comfortable merging this for a 0.13 release (at least not yet). We will hold off until we can get some real-world tests in an 0.14 alpha release, and if all goes well in 0.14 prereleases we will backport this fix to 0.13.

**Comment by apparentlymart** [comment]
I think we should be conservative and avoid backporting to 0.13 anything that involves upgrading `cty` past its 1.6.0 release, because that release includes bug fixes that change behavior to be more correct (per the design) but may produce subtly different results in some edge cases.

If we end up with something sufficiently critical that needs to be backported then I (with my `cty` maintainer hat on) would be willing to backport certain `cty` changes to a 1.5.x patch release, but I don't think this one meets that bar because it's essentially just replacing one error message with another error message: the `cty` function machinery was already catching the panic and returning it as a normal error. Although the message includes a panic trace this isn't actually a "crash" in the way we usually think of it, because Terraform is still able to run to completion, clean up its work in progress, and return the error message.

**Comment by mildwonkey** [comment]
👍 I am going to close this issue then - it's fixed for 0.14, and it doesn't appear likely we'll be backporting to 0.13. Thanks!
